Data.Rmd

```diff
@@ -0,0 +1,462 @@
+  }
+}
+```
+
+```{r}
+readxl::readxl_example()
+
+readxl::readxl_example("clippy.xlsx")
+```
+
+## Internal state
+
+Sometimes there's information that multiple functions from your package need to access that:
+
+-   Must be determined at runtime, not build time. It might even be dynamic.
+-   Doesn't make sense to pass in via a function argument. Often it's some obscure detail that a user shouldn't even know about.
+
+A great way to manage such data is to use an *environment*.[^data-1]
+This environment must be created a build time, but you can populate it with values at runtime and update those values over the course of an R session.
+This works because environments have reference semantics (whereas more pedestrian R objects, such as atomic vectors, lists, or data frames have value semantics).
+
+[^data-1]: If you don't know much about R environments and what makes them special, a great resource is the [Environments chapter](https://adv-r.hadley.nz/environments.html) of Advanced R.
+
+Consider a package that can store the user's favorite letters or numbers.
+You might start out with code like this in a file below `R/`:
+
+```{r eval = FALSE}
+favorite_letters <- letters[1:3]
+
+#' Report my favorite letters
+#' @export
+mfl <- function() {
+  favorite_letters
+}
+
+#' Change my favorite letters
+#' @export
+set_mfl <- function(l = letters[24:26]) {
+  favorite_letters <<- l
+  favorite_letters
+}
+```
+
+`favorite_letters` is initialized to ("a", "b", "c") when the package is built.
+The user can then inspect `favorite_letters` with `mfl()`, at which point they'll probably want to register *their* favorite letters with `set_mfl()`.
+Note that we've used the super assignment operator `<<-` in `set_mfl()` in the hope that this will reach up into the package environment and modify the internal data object `favorite_letters`.
+But a call to `set_mfl()` fails like so:
+
+```{r, eval = FALSE}
+set_mfl(c("j", "f", "b"))
+#> Error in set_mfl(c("j", "f", "b")): cannot change value of locked binding for
+#> 'favorite_letters'
+```
+
+Because `favorite_letters` is a regular character vector, modification requires making a copy and rebinding the name `favorite_letters` to this new value.
+And that is what's disallowed: you can't change the binding for objects in the package namespace (well, at least not without trying harder than this).
+
+However, if we maintain our internal state within an internal package environment, we can modify objects contained in the environment and even add completely new objects.
+Here's an alternative implementation that uses an internal environment named `the`.
+
+```{r, eval = FALSE}
+the <- new.env(parent = emptyenv())
+the$favorite_letters <- letters[1:3]
+
+#' Report my favorite letters
+#' @export
+mfl2 <- function() {
+  the$favorite_letters
+}
+
+#' Change my favorite letters
+#' @export
+set_mfl2 <- function(l = letters[24:26]) {
+  the$favorite_letters <- l
```

Comment:
If you're going to use a set function as an example, I think you should follow our standard practice and invisibly return the previous value.

Reply:
done